src/store/hooks/useAuth.tsx

```diff
@@ -0,0 +1,223 @@
+import {
+  createContext, useContext, useMemo, ReactNode,
+  useEffect, useState,
+} from 'react';
+import {
+  getAuth, onAuthStateChanged, User, signOut, Auth,
+} from 'firebase/auth';
+import { LoadingOverlay } from '@mantine/core';
+import { useStorageEngine } from '../storageEngineHooks';
+import { FirebaseStorageEngine } from '../../storage/engines/FirebaseStorageEngine';
+import { GlobalConfig } from '../../parser/types';
+
+interface LocalStorageUser{
+  name: string,
+  email: string
+}
+
+interface UserWrapped{
+  user:User|LocalStorageUser|null,
+  determiningStatus:boolean,
+  isAdmin:boolean
+}
+
+type verifyAdminStatusInput = User | null | LocalStorageUser;
+
+// Defines default AuthContextValue
+interface AuthContextValue {
+  user: UserWrapped;
+  adminVerification:boolean;
+  login: (user: UserWrapped) => void;
+  logout: () => void;
+  verifyAdminStatus: (firebaseUser:verifyAdminStatusInput) => Promise<boolean>;
+  }
+
+// Initializes AuthContext
+const AuthContext = createContext<AuthContextValue>({
+  user: {
+    user: null,
+    determiningStatus: false,
+    isAdmin: false,
+  },
+  adminVerification: false,
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  login: () => {},
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  logout: () => {},
+  verifyAdminStatus: () => Promise.resolve(false),
+
+});
+
+// Firebase auth context
+export const useAuth = () => useContext(AuthContext);
+
+// Defines the functions that are exposed in this hook.
+export function AuthProvider({ children, globalConfig } : { children: ReactNode, globalConfig: GlobalConfig }) {
+  const [user, setUser] = useState<UserWrapped>(
+    {
+      user: null,
+      determiningStatus: true,
+      isAdmin: false,
+    },
+  );
+
+  const [adminVerification, setAdminVerification] = useState<boolean>(false);
+
+  const { storageEngine } = useStorageEngine();
+
+  // Logs in the user (i.e. sets the user info and navigates to the root)
+  const login = (inputUser: UserWrapped) => {
+    setUser(inputUser);
+  };
```

Comment:
You'll need to play with this, but this should allow you to await the login. You'll need something similar for logout to await that too

```suggestion
  const login = async (inputUser: UserWrapped) => {
    return new Promise((resolve, reject) => {
      setUser(inputUser, () => resolve());
    }
  };
```

Reply:
Ahhh, this makes a lot of sense. Thank you!!